plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -0,0 +1,228 @@
+        ctx.put("_parent", oldParent);
+        String oldRouting = fieldValue(doc, "_routing");
+        ctx.put("_routing", oldRouting);
+        Long oldTimestamp = fieldValue(doc, "_timestamp");
+        ctx.put("_timestamp", oldTimestamp);
+        Long oldTtl = fieldValue(doc, "_ttl");
+        ctx.put("_ttl", oldTtl);
+        ctx.put("_source", index.sourceAsMap());
+        ctx.put("op", "update");
+        script.setNextVar("ctx", ctx);
+        script.run();
+        Map<String, Object> resultCtx = (Map<String, Object>) script.unwrap(ctx);
+        String newOp = (String) resultCtx.remove("op");
+        if (newOp == null) {
+            throw new IllegalArgumentException("Script cleared op!");
+        }
+        if ("noop".equals(newOp)) {
+            noops.incrementAndGet();
+            return false;
+        }
+        if ("update".equals(newOp) == false) {
+            throw new IllegalArgumentException("Invalid op [" + newOp + ']');
+        }
+
+        /*
+         * It'd be lovely to only set the source if we know its been
+         * modified but it isn't worth keeping two copies of it
+         * around just to check!
+         */
+        index.source((Map<String, Object>) resultCtx.remove("_source"));
+
+        Object newValue = ctx.remove("_index");
+        if (false == doc.index().equals(newValue)) {
+            scriptChangedIndex(index, newValue);
+        }
+        newValue = ctx.remove("_type");
+        if (false == doc.type().equals(newValue)) {
+            scriptChangedType(index, newValue);
+        }
+        newValue = ctx.remove("_id");
+        if (false == doc.id().equals(newValue)) {
+            scriptChangedId(index, newValue);
+        }
+        newValue = ctx.remove("_version");
+        if (false == Objects.equals(oldVersion, newValue)) {
+            scriptChangedVersion(index, newValue);
+        }
+        newValue = ctx.remove("_parent");
+        if (false == Objects.equals(oldParent, newValue)) {
+            scriptChangedParent(index, newValue);
+        }
+        /*
+         * Its important that routing comes after parent in case you want to
+         * change them both.
+         */
+        newValue = ctx.remove("_routing");
+        if (false == Objects.equals(oldRouting, newValue)) {
+            scriptChangedRouting(index, newValue);
+        }
+        newValue = ctx.remove("_timestamp");
+        if (false == Objects.equals(oldTimestamp, newValue)) {
+            scriptChangedTimestamp(index, newValue);
+        }
+        newValue = ctx.remove("_ttl");
+        if (false == Objects.equals(oldTtl, newValue)) {
+            scriptChangedTtl(index, newValue);
+        }
+        if (false == ctx.isEmpty()) {
+            throw new IllegalArgumentException("Invalid fields added to ctx [" + String.join(",", ctx.keySet()) + ']');
+        }
+        return true;
+    }
+
+    protected void scriptChangedIndex(IndexRequest index, Object to) {
```

Comment:
It seems to me that these defaults are good for one subclass but not for the other. We could just make these methods abstract here and implement them in both subclasses?